include/tvm/diagnostic/GraphProbe.h

```diff
@@ -0,0 +1,270 @@
+   * for this base class.
+   */
+#ifndef _MSC_VER
+  template<typename T, typename U, typename EnumOutput, typename Base = T>
+  void registerAccessor(
+      EnumOutput o,
+      const U & (Base::*fn)() const,
+      std::function<Eigen::MatrixXd(const U &)> convert = [](const U & u) { return u; });
+#else
+  template<typename T, typename U, typename EnumOutput>
+  void registerAccessor(
+      EnumOutput o,
+      const U & (T::*fn)() const,
+      std::function<Eigen::MatrixXd(const U &)> convert = [](const U & u) { return u; });
+#endif
+
+  /** Register a method \p fn taking a variable to retrieve the value associated to \p o
+   *
+   * \tparam T type of the node to which the method is attached. It must be given explicitly
+   * if the method is inherited from a base class, otherwise the method will be registered
+   * for this base class.
+   */
+#ifndef _MSC_VER
+  template<typename T, typename U, typename EnumOutput, typename Base = T>
+  void registerAccessor(
+      EnumOutput o,
+      U (Base::*fn)(const Variable &) const,
+      std::function<Eigen::MatrixXd(const U &)> convert = [](const U & u) { return u; });
+#else
+  template<typename T, typename U, typename EnumOutput>
+  void registerAccessor(
+      EnumOutput o,
+      U (T::*fn)(const Variable &) const,
+      std::function<Eigen::MatrixXd(const U &)> convert = [](const U & u) { return u; });
+#endif
+
+  /** Register all methods associated to outputs inherited from tvm::function::abstract::Function */
+  template<typename T>
+  void registerTVMFunction();
+
+  /** Register all methods associated to outputs inherited from tvm::constraint::abstract::Constraint */
+  template<typename T>
+  void registerTVMConstraint();
+
+  /** Register all methods associated to outputs inherited from tvm::task_dynamics::abstract::TaskDynamicsImpl */
+  template<typename T>
+  void registerTVMTaskDynamics();
+
+  /** List all the values of all the outputs present in the call graph \p g (if the associated
+   * methods were registered).
+   *
+   * If \p verbose is \c true, display the outputs for which no methods were registered to retrieve
+   * their values.
+   */
+  std::vector<OutputVal> listOutputVal(const graph::CallGraph * const g, bool verbose = false) const;
+
+  /** List all the values of all the outputs on which \p o depends (recursively) (if the associated
+   * methods were registered).
+   *
+   * If \p verbose is \c true, display the outputs for which no methods were registered to retrieve
+   * their values.
+   */
+  std::vector<OutputVal> listOutputVal(const Output & o, bool verbose = false) const;
+
+  /** Follow up backward the computation graph starting at node \p o, and trimming out branches where the
+   * value of an output does not pass the test given by \p select.
+   */
+  std::unique_ptr<Node> followUp(
+      const Output & o,
+      std::function<bool(const Eigen::MatrixXd &)> select = [](const Eigen::MatrixXd &) { return true; }) const;
```

Comment:
Should be `const std::function<bool(const Eigen::MatrixXd &)> & select` to avoid potentially expensive copies (same with the internal functions)

Reply:
Fixed.